experimental/sql/src/projection_optimizations.py

```diff
@@ -0,0 +1,191 @@
+  # tables) that map away all columns that aren't needed by this Operator or
+  # further up the tree.
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Operator, rewriter: PatternRewriter):
+    if isinstance(op, RelAlg.Project) or isinstance(
+        op, RelAlg.Table) or isinstance(op, RelAlg.SchemaElement):
+      return
+    cols = list(dict.fromkeys(self.find_cols_upstream(op)))
+    if cols.__contains__(None):
+      return
+
+    new_proj = RelAlg.Project.get(
+        rewriter.move_region_contents_to_new_regions(op.input),
+        Region.from_operation_list([RelAlg.Column.get(s) for s in cols]),
+        ArrayAttr.from_list([StringAttr.from_str(s) for s in cols]))
+
+    # TODO: I mainly need this pattern match to call the proper builder. This
+    # could maybe be done nicer.
+    if isinstance(op, RelAlg.Select):
+      rewriter.replace_matched_op(
+          RelAlg.Select.get(
+              Region.from_operation_list([new_proj]),
+              rewriter.move_region_contents_to_new_regions(op.predicates)))
+    elif isinstance(op, RelAlg.Aggregate):
+      rewriter.replace_matched_op(
+          RelAlg.Aggregate.get(Region.from_operation_list([new_proj]),
+                               [s.data for s in op.col_names.data],
+                               [s.data for s in op.functions.data],
+                               [s.data for s in op.res_names.data]))
+
+
+@dataclass
+class ProjectionFuser(ProjectionOptimizer):
+
+  # This rewriter fuses two subsequent projections. Currently, it only works on
+  # renaming style projections, but it will be extended in the future.
+
+  def rename_expr(self, op: RelAlg.Expression, map: dict[str,
+                                                         str]) -> Operation:
+    if isinstance(op, RelAlg.Column):
+      return RelAlg.Column.get(map[op.col_name.data])
+    if isinstance(op, RelAlg.Multiply):
+      return RelAlg.Multiply.get(
+          Region.from_operation_list([self.rename_expr(op.lhs, map)]),
+          Region.from_operation_list([self.rename_expr(op.rhs, map)]))
+    if isinstance(op, RelAlg.Literal):
+      return RelAlg.Literal.get(op.val, op.type)
+    if isinstance(op, RelAlg.Compare):
+      return RelAlg.Multiply.get(
+          op.comparator.data,
+          Region.from_operation_list([self.rename_expr(op.left, map)]),
+          Region.from_operation_list([self.rename_expr(op.right, map)]))
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Project, rewriter: PatternRewriter):
+    if not isinstance(op.input.op, RelAlg.Project):
+      return
+
+    child_cols = self.flatten(
+        [self.find_cols_in_expr(o) for o in op.input.op.projections.ops])
+    child_res = [s.data for s in op.input.op.names.data]
+    # If the child projection simply projects onto a subset, we can just infer
+    # its column mapping and apply it on the parents projection.
+    if len(child_cols) == len(child_res) and all(
```

Comment:
The code rather seems to test for a rename but the comment talks about "subset". Is the comment out of date?